src/runtime/patch/fs/fs.js

```diff
@@ -0,0 +1,356 @@
+      return false;
+    }
+
+    fs.getValidatedFD(fd);
+    return existsInFs(fs, fd.filePath);
+  } catch {
+    /* no-op */
+  }
+  return false;
+}
+
+function isOnFs(fs, path, fd = null) {
+  return existsInFs(fs, path) || existsFdInFs(fs, fd);
+}
+
+// Node Fs module patched methods
+function close(sfs, realFs, fd, callback) {
+  if (isOnFs(sfs, null, fd)) {
+    return sfs.close(fd, callback);
+  }
+
+  return realFs.close(fd, callback);
+}
+
+function createReadStream(sfs, realFs, path, options) {
+  const optionsFd = options && options.fd;
+  const isOnRealFs = isOnFs(realFs, path, optionsFd);
+  const isOnSfs = isOnFs(sfs, path, optionsFd);
+
+  if (isOnSfs && !isOnRealFs) {
+    return sfs.createReadStream(path, options);
+  }
+
+  return realFs.createReadStream(path, options);
+}
+
+function exists(sfs, realFs, path, callback) {
+  const isOnRealFs = isOnFs(realFs, path);
+  const isOnSfs = isOnFs(sfs, path);
+
+  if (isOnSfs || isOnRealFs) {
+    return callback(true);
+  }
+
+  return callback(false);
+}
+
+function fstat(sfs, realFs, fd, options, callback) {
+  const sanitizedOptions = typeof options === 'function' ? { bigint: false } : options;
+  const sanitizedCallback = typeof callback === 'function' ? callback : options;
+  const isOnSfs = isOnFs(sfs, null, fd);
+
+  if (isOnSfs) {
+    return sfs.fstat(fd, sanitizedOptions, sanitizedCallback);
+  }
+
+  return realFs.fstat(fd, sanitizedOptions, sanitizedCallback);
+}
+
+function open(sfs, realFs, path, flags, mode, callback) {
+  let sanitizedCallback;
+  let sanitizedMode;
+  let sanitizedFlags;
+  const isOnRealFs = isOnFs(realFs, path);
+  const isOnSfs = isOnFs(sfs, path);
+
+  if (typeof mode === 'function') {
+    sanitizedCallback = mode;
+    sanitizedMode = 0o666;
+    sanitizedFlags = flags;
+  } else if (typeof flags === 'function') {
+    sanitizedCallback = flags;
+    sanitizedMode = 0o666;
+    sanitizedFlags = 'r';
+  } else {
+    sanitizedCallback = callback;
+    sanitizedMode = mode;
+    sanitizedFlags = flags;
+  }
+
+  if (isOnSfs && !isOnRealFs) {
+    return sfs.open(path, sanitizedCallback);
+  }
+
+  return realFs.open(path, sanitizedFlags, sanitizedMode, sanitizedCallback);
+}
+
+function read(sfs, realFs, fd, buffer, offset, length, position, callback) {
+  const isOnRealFs = isOnFs(realFs, null, fd);
+  const isOnSfs = isOnFs(sfs, null, fd);
+
+  if (isOnSfs && !isOnRealFs) {
+    return sfs.read(fd, buffer, offset, length, position, callback);
+  }
+
+  return realFs.read(fd, buffer, offset, length, position, callback);
+}
+
+function readdir(sfs, realFs, path, options, callback) {
+  const sanitizedCallback = typeof callback === 'function' ? callback : options;
+  const sanitizedOptions = typeof options === 'object' ? options : { encoding: 'utf8', withFileTypes: false };
+  const isOnRealFs = isOnFs(realFs, path);
+  const isOnSfs = isOnFs(sfs, path);
+
+  if (isOnRealFs && isOnSfs) {
+    const sfsReadDir = new Promise((resolve, reject) => {
```

Comment:
**question** can you give an example of when we'd want to read the contents of both a real and static FS directory? This seems inconsistent with `readFile`, where we prefer the contents of the real FS over the contents of the static FS

Reply:
yeah I can. It is the only outsider for the common logic.
Imagine for example that we build a static-fs volume for the entire node_modules we have on Kibana. 

After the build one of the added folders to the static-fs volume would be `x-pack/plugins/infra/node_modules` and will not exist in the realFS anymore. However `x-pack/plugins/infra` will keep existing in the realFS but at the same time will also exist in the static-fs.

If you proceed with a fs.readdir on `x-pack/plugins/infra` the expected result is what we have both on the static-fs and into the realfs (which is happening in the platform logic to startup plugins).